pkg/kubelet/cm/dra/plugin/client.go

```diff
@@ -150,57 +103,82 @@ func (r *draPluginClient) NodeUnprepareResources(
 
-	if r.nodeClientCreator == nil {
-		return nil, errors.New("failed to call NodeUnprepareResources. nodeClientCreator is nil")
-	}
-
-	nodeClient, nodeClientOld, closer, err := r.nodeClientCreator(r.addr)
+	conn, err := r.plugin.getOrCreateGRPCConn()
 	if err != nil {
 		return nil, err
 	}
-	defer closer.Close()
+	nodeClient := drapb.NewNodeClient(conn)
 
 	ctx, cancel := context.WithTimeout(ctx, PluginClientTimeout)
 	defer cancel()
 
+	if r.plugin.getHighestSupportedGRPCApi() != highestAvailableGRPCApi {
+		// we already tried the v1alpha3 API but it isn't available so falling back automatically
+		resp = nodeUnrepareResourceFallback(ctx, conn, req)
+		return
+	}
+
 	resp, err = nodeClient.NodeUnprepareResources(ctx, req)
 	if err != nil {
 		status, _ := grpcstatus.FromError(err)
 		if status.Code() == grpccodes.Unimplemented {
 			// Fall back to the older gRPC API.
-			resp = &drapb.NodeUnprepareResourcesResponse{
-				Claims: make(map[string]*drapb.NodeUnprepareResourceResponse),
-			}
-			err = nil
-			for _, claim := range req.Claims {
-				_, errOld := nodeClientOld.NodeUnprepareResource(ctx,
-					&drapbv1alpha2.NodeUnprepareResourceRequest{
-						Namespace:      claim.Namespace,
-						ClaimUid:       claim.Uid,
-						ClaimName:      claim.Name,
-						ResourceHandle: claim.ResourceHandle,
-					})
-				result := &drapb.NodeUnprepareResourceResponse{}
-				if errOld != nil {
-					result.Error = errOld.Error()
-				}
-				resp.Claims[claim.Uid] = result
-			}
+			r.plugin.setHighestSupportedGRPCApi("v1alpha2")
+			resp = nodeUnrepareResourceFallback(ctx, conn, req)
 		}
 	}
 
 	return
 }
 
-func newGrpcConn(addr draAddr) (*grpc.ClientConn, error) {
-	network := "unix"
-	klog.V(4).InfoS(log("creating new gRPC connection"), "protocol", network, "endpoint", addr)
-
-	return grpc.Dial(
-		string(addr),
-		grpc.WithTransportCredentials(insecure.NewCredentials()),
-		grpc.WithContextDialer(func(ctx context.Context, target string) (net.Conn, error) {
-			return (&net.Dialer{}).DialContext(ctx, network, target)
-		}),
-	)
+func nodePrepareResourceFallback(ctx context.Context, conn *grpc.ClientConn, req *drapb.NodePrepareResourcesRequest) *drapb.NodePrepareResourcesResponse {
+	nodeClient := drapbv1alpha2.NewNodeClient(conn)
+	response := &drapb.NodePrepareResourcesResponse{
+		Claims: make(map[string]*drapb.NodePrepareResourceResponse),
+	}
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	for _, claim := range req.Claims {
+		res, err := nodeClient.NodePrepareResource(ctx,
+			&drapbv1alpha2.NodePrepareResourceRequest{
+				Namespace:      claim.Namespace,
+				ClaimUid:       claim.Uid,
+				ClaimName:      claim.Name,
+				ResourceHandle: claim.ResourceHandle,
+			})
+		result := &drapb.NodePrepareResourceResponse{}
+		if err != nil {
+			result.Error = err.Error()
+		} else {
+			result.CDIDevices = res.CdiDevices
+		}
+		response.Claims[claim.Uid] = result
+	}
+	return response
+}
```

Comment:
I'd rather see explicit functions for calling the v1alpha2 vs. the v1alpha3 API.

I.e. something like the following:
1. Define a new type to store the detected API version.  Return this from the call to `detectVersion()` (or whatever we want to call it) mentioned earlier.
```
type versionedClient struct {
    version string
}
```
3. Hang functions off of this type for each supported API
```
func (v *versionedClient)  v1alpha2NodePrepareResource(ctx context.Context, conn *grpc.ClientConn, req *drapbv1alpha2.NodePrepareResourceRequest) *drapbv1alpha2.NodePrepareResourceResponse {
...
}

func (v *versionedClient)  v1alpha3NodePrepareResources(ctx context.Context, conn *grpc.ClientConn, req *drapbv1alpha2.NodePrepareResourcesRequest) *drapbv1alpha2.NodePrepareResourcesResponse {
...
}

func (v *versionedClient) v1alpha2NodeUnrepareResource(ctx context.Context, conn *grpc.ClientConn, req *drapbv1alpha2.NodeUnprepareResourceRequest) *drapbv1alpha2.NodeUnprepareResourceResponse {
...
}

func (v *versionedClient) v1alpha3NodeUnrepareResources(ctx context.Context, conn *grpc.ClientConn, req *drapbv1alpha3.NodeUnprepareResourcesRequest) *drapbv1alpha3.NodeUnprepareResourcesResponse {
...
}
```
4. Write explicit functions hanging off of `versionedClient` to do any necessary conversions, e.g v1alpha3 --> v1alpha2 PrepareRequest, and v1alpha2 --> v1alpha3 PrepareResponse


5. In the bodies of `draPluginClient.NodePrepareResources()` and `draPluginClient.NodeUnprepareResources()` check `versionedAPIs.version` to see which API needs to be called, and then perform a series of explicit calls to the functions hanging off of `versionedClient` to make the necessary calls / conversions.

Reply:
done in 832264ac0d4